### PR TITLE
feat: New resource/datasource `cloudavenue_backup` 

### PR DESCRIPTION
**Pull request by dmicheneau**
Add new resource cloudavenue_backup

If you submit change in the provider code, please make sure to:

- [x] Write or modify examples in `examples/` directory
- [x] Write or modify acceptance tests
- [x] Run `make generate` to ensure the doc was updated properly

### How has this code been tested
 
```
source ~/.cloud_avenue.source && TF_ACC=1 go test  -v ./internal/testsacc/ -count=1 -timeout 30m -run TestAccBackupResource
=== RUN   TestAccBackupResource
--- PASS: TestAccBackupResource (238.81s)
PASS
ok      github.com/orange-cloudavenue/terraform-provider-cloudavenue/internal/testsacc  239.111s
```

```
source ~/.cloud_avenue.source && TF_ACC=1 go test -v ./internal/testsacc/ -count=1 -timeout 30m -run TestAccBackupDataSource
=== RUN   TestAccBackupDataSource
--- PASS: TestAccBackupDataSource (176.99s)
PASS
ok      github.com/orange-cloudavenue/terraform-provider-cloudavenue/internal/testsacc  177.284s
```

### CODE REVIEW

**Comment by dmicheneau** [comment]
Waiting to solve WAF issue in order to implement refresh sync between NetBackup and VCD.
And also will permit to Implement TestACC

File: internal/provider/backup/backup_schema.go [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
+
+	"github.com/hashicorp/terraform-plugin-framework-validators/setvalidator"
+	"github.com/hashicorp/terraform-plugin-framework-validators/stringvalidator"
+
+	superschema "github.com/FrangipaneTeam/terraform-plugin-framework-superschema"
+)
+
+// TODO : Remove unused imports.
```
Comment (azrod) [inline internal/provider/backup/backup_schema.go:22]: Remove unused imports

File: internal/provider/backup/backup_schema.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+						stringplanmodifier.RequiresReplace(),
+					},
+				},
+				DataSource: &schemaD.StringAttribute{
+					Computed: true,
+				},
+			},
+			"target_id": superschema.SuperStringAttribute{
+				Common: &schemaR.StringAttribute{
+					MarkdownDescription: "The ID of the target. A target can be a VDC, a VApp or a VM.",
```
Comment (azrod) [inline internal/provider/backup/backup_schema.go:117]: ID is URN vmware format ?

File: internal/provider/backup/backup_types.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+func (rm *backupModel) getPolicies(ctx context.Context) (values *backupModelPolicies, diags diag.Diagnostics) {
+	values = &backupModelPolicies{}
+	d := rm.Policies.Get(ctx, &values, false)
+	return values, d
+}
+
+// SetPolicies sets the value of the Policies field.
+func (rm *backupModel) setPolicies(ctx context.Context, values *backupModelPolicies) diag.Diagnostics {
+	return rm.Policies.Set(ctx, values)
+}
```
Comment (azrod) [inline internal/provider/backup/backup_types.go:60]: Why ?

File: docs/resources/backup.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
+---
+page_title: "cloudavenue_backup Resource - cloudavenue"
+subcategory: "Backup"
+description: |-
+  The cloudavenue_backup resource allows you to manage backup from NetBackup solution.
+---
+
+# cloudavenue_backup (Resource)
+
+The `cloudavenue_backup` resource allows you to manage backup from NetBackup solution.
```
Comment (azrod) [inline docs/resources/backup.md:10]: Add more detail : Backup is available for VDC, VAPP or VM and add link to documentation

**Review by azrod — CHANGES_REQUESTED** [review]
Add Datasource

File: internal/provider/backup/backup_resource.go [hunk 5]
```diff
@@ -0,0 +1,456 @@
...
+
+// Unapply the protection level for a policy to the target.
+// A target can be a VDC, VAPP or VM.
+// Return an error if any.
+func unApplyPolicy[T target](t T, policy backupModelPolicy) error {
+	var job *common.JobAPIResponse
+	var err error
+
+	// apply the protection levels
+	job, err = t.Unprotect(netbackupclient.ProtectUnprotectRequest{
```
[the remark this answers is not in the document]
Comment (dmicheneau) [reply internal/provider/backup/backup_resource.go:375]: Fix

File: go.mod [hunk 6]
```diff
@@ -104,3 +104,5 @@ require (
 	google.golang.org/grpc v1.57.0 // indirect
 	google.golang.org/protobuf v1.31.0 // indirect
 )
+
+replace github.com/orange-cloudavenue/netbackup-sdk-go => /Users/micheneaudavid/go/src/github.com/orange-cloudavenue/netbackup-sdk-go
```
[the remark this answers is not in the document]
Comment (dmicheneau) [reply go.mod:108]: Fix

File: internal/provider/backup/backup_schema.go (hunk 1 above)
Comment (dmicheneau) [reply internal/provider/backup/backup_schema.go:22]: Fix

File: internal/provider/backup/backup_types.go (hunk 3 above)
Comment (dmicheneau) [reply internal/provider/backup/backup_types.go:60]: Fix

File: internal/provider/backup/backup_schema.go (hunk 2 above)
Comment (dmicheneau) [reply internal/provider/backup/backup_schema.go:66]: Fix

File: templates/resources/backup.md.tmpl [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+resource "cloudavenue_backup" "example-vdc" {
+  type = "vdc"
+  target_name = cloudavenue_vdc.example.name
+  policies = [{
+      policy_name = "D6"
+    } ,{
+      policy_name = "M3"
+     }
+  ]
+}
```
Comment (azrod) [inline templates/resources/backup.md.tmpl:24]:
```suggestion
resource "cloudavenue_backup" "example-vdc" {
  type = "vdc"
  target_name = cloudavenue_vdc.example.name
  policies = [{
      policy_name = "D6"
    } ,
    {
      policy_name = "M3"
     }
  ]
}
```

File: templates/resources/backup.md.tmpl [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+      policy_name = "D6"
+    }]
+}
+```
+
+### Example Usage of a VM Backup with a policy set
+```hcl
+resource "cloudavenue_backup" "example-vm" {
+  type = "vm"
+  target_name = "vmdemo"
```
Comment (azrod) [inline templates/resources/backup.md.tmpl:42]:
```suggestion
  target_name = cloudavenue_vm.example.name
```

File: internal/provider/backup/backup_schema.go [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+			},
+			"type": superschema.SuperStringAttribute{
+				Common: &schemaR.StringAttribute{
+					MarkdownDescription: "Scope of the backup.",
+					Required:            true,
+				},
+				Resource: &schemaR.StringAttribute{
+					Validators: []validator.String{
+						stringvalidator.OneOf("vdc", "vapp", "vm"),
+					},
```
Comment (azrod) [inline internal/provider/backup/backup_schema.go:55]: Move this to common

File: internal/provider/backup/backup_schema.go [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+				},
+			},
+			"target_id": superschema.SuperStringAttribute{
+				Common: &schemaR.StringAttribute{
+					MarkdownDescription: "The ID of the target. A target can be a VDC, a VApp or a VM.",
+					Optional:            true,
+				},
+				Resource: &schemaR.StringAttribute{
+					Validators: []validator.String{
```
Comment (azrod) [inline internal/provider/backup/backup_schema.go:67]: Move this to common

File: internal/provider/backup/backup_schema.go [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+					Computed: true,
+				},
+			},
+			"target_name": superschema.SuperStringAttribute{
+				Common: &schemaR.StringAttribute{
+					MarkdownDescription: "The name of the target. A target can be a VDC, a VApp or a VM.",
+					Optional:            true,
+				},
+				Resource: &schemaR.StringAttribute{
+					Validators: []validator.String{
```
Comment (azrod) [inline internal/provider/backup/backup_schema.go:86]: move this to common

File: internal/provider/backup/backup_schema.go [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+					"policy_name": superschema.SuperStringAttribute{
+						Common: &schemaR.StringAttribute{
+							MarkdownDescription: "The name of the backup policy.",
+						},
+						Resource: &schemaR.StringAttribute{
+							Required: true,
+							Validators: []validator.String{
+								stringvalidator.OneOf("D6", "D30", "D30NQ", "D60", "W4", "M3", "M12", "XD6", "XD30", "XD60", "XW4", "XM3", "XM12"),
+							},
+							MarkdownDescription: "Each letter represent a strategy predefined: D = Daily, W = Weekly, M = Monthly, X = Replication, The number is the retention period. Please refer to the documentation for more information. https://wiki.cloudavenue.orange-business.com/wiki/Backup",
```
Comment (azrod) [inline internal/provider/backup/backup_schema.go:127]: Move this to common markdown description and change link to markdown format https://www.markdownguide.org/basic-syntax/#links

File: internal/provider/backup/backup_schema_test.go [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+	}
+
+	// Validate the schema
+	diagnostics := schemaResponse.Schema.ValidateImplementation(ctx)
+
+	if diagnostics.HasError() {
+		t.Fatalf("Schema validation diagnostics: %+v", diagnostics)
+	}
+}
+*/
```
Comment (azrod) [inline internal/provider/backup/backup_schema_test.go:61]: Uncomment this

File: internal/provider/provider_datasources.go [hunk 14]
```diff
@@ -6,6 +6,7 @@ import (
 	"github.com/hashicorp/terraform-plugin-framework/datasource"
 
 	"github.com/orange-cloudavenue/terraform-provider-cloudavenue/internal/provider/alb"
+	netbackup "github.com/orange-cloudavenue/terraform-provider-cloudavenue/internal/provider/backup"
```
Comment (azrod) [inline internal/provider/provider_datasources.go:9]: Why renaming backup to netbackup ?

File: internal/testsacc/backup_datasource_test.go
Comment (azrod) [inline internal/testsacc/backup_datasource_test.go:1]: Add test datasource with target_id

File: internal/testsacc/backup_resource_test.go
Comment (azrod) [inline internal/testsacc/backup_resource_test.go:1]: Add test with target_id

**Review by azrod — CHANGES_REQUESTED** [review]

File: docs/resources/backup.md [hunk 15]
```diff
@@ -0,0 +1,82 @@
+---
+page_title: "cloudavenue_backup Resource - cloudavenue"
+subcategory: "Backup"
+description: |-
+  The cloudavenue_backup resource allows you to manage backup strategy for vdc,vapp and 'vm' from NetBackup solution. Please refer to the documentation for more information. https://wiki.cloudavenue.orange-business.com/wiki/Backup
+---
+
+# cloudavenue_backup (Resource)
+
+The `cloudavenue_backup` resource allows you to manage backup strategy for `vdc`,`vapp` and 'vm' from NetBackup solution. Please refer to the documentation for more information. https://wiki.cloudavenue.orange-business.com/wiki/Backup
```
[the remark this answers is not in the document]
Comment (dmicheneau) [reply docs/resources/backup.md:10]: Fix

File: internal/provider/backup/backup_resource.go [hunk 16]
```diff
@@ -0,0 +1,475 @@
...
+	// get the protection level ID
+	if policy.PolicyID.Get() == 0 {
+		pl, err := t.GetProtectionLevelAvailableByName(policy.PolicyName.Get())
+		if err != nil {
+			return backupModelPolicy{}, err
+		}
+		policy.PolicyID.SetInt(pl.ID)
+	}
+	return policy, nil
+}
```
[the remark this answers is not in the document]
Comment (dmicheneau) [reply internal/provider/backup/backup_resource.go:360]: Fix

File: internal/provider/backup/backup_resource.go [hunk 17]
```diff
@@ -0,0 +1,475 @@
...
+// Return an error if any.
+func unApplyPolicies[T target](t T, policies *backupModelPolicies) (err error) {
+	for _, policy := range *policies {
+		err := unApplyPolicy(t, policy)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (dmicheneau) [reply internal/provider/backup/backup_resource.go:387]: Fix

File: internal/provider/backup/backup_schema.go (hunk 12 above)
Comment (dmicheneau) [reply internal/provider/backup/backup_schema.go:127]: Fix

File: internal/provider/backup/backup_schema_test.go (hunk 13 above)
Comment (dmicheneau) [reply internal/provider/backup/backup_schema_test.go:61]: Fix

File: internal/provider/provider_datasources.go (hunk 14 above)
Comment (dmicheneau) [reply internal/provider/provider_datasources.go:9]: no reason, Fix

File: internal/testsacc/backup_datasource_test.go
Comment (dmicheneau) [reply internal/testsacc/backup_datasource_test.go:1]: fix

File: internal/testsacc/backup_resource_test.go
Comment (dmicheneau) [reply internal/testsacc/backup_resource_test.go:1]: fix in example 2

File: internal/client/client.go [hunk 18]
```diff
@@ -77,9 +80,10 @@ func (c *CloudAvenue) New() (*CloudAvenue, error) {
 	}
 
 	// API NetBackup
-	if c.NetBackup.IsDefined() {
-		if err := c.NewNetBackupClient(); err != nil {
-			return nil, err
+	if c.NetBackupOpts.Netbackup.Endpoint != "" && c.NetBackupOpts.Netbackup.Username != "" && c.NetBackupOpts.Netbackup.Password != "" {
```
Comment (azrod) [inline internal/client/client.go:83]: Endpoint is not mandatory

File: internal/provider/provider.go [hunk 19]
```diff
@@ -106,30 +111,29 @@ func (p *cloudavenueProvider) Configure(ctx context.Context, req provider.Config
...
-				"Missing NetBackup API Password",
-				"The provider cannot create the NetBackup API client as there is a missing or empty value for the NetBackup API password. "+
-					"Set the host value in the configuration or use the NETBACKUP_PASSWORD environment variable. "+
-					"If either is already set, ensure the value is not empty.",
-			)
-		}
+	// Default URL to the public NetBackup API if not set.
+	if cloudAvenue.NetBackupOpts.Netbackup.Endpoint == "" {
+		cloudAvenue.NetBackupOpts.Netbackup.Endpoint = "https://backup1.cloudavenue.orange-business.com/NetBackupSelfServiceNetBackupPanels/Api"
+	}
```
Comment (azrod) [inline internal/provider/provider.go:117]: This is managed by the SDK

File: internal/provider/backup/backup_resource.go [hunk 20]
```diff
@@ -0,0 +1,475 @@
...
+	}
+
+	return nil
+}
+
+// read is a generic read function that can be used by the resource Create, Read and Update functions.
+func (r *backupResource) read(ctx context.Context, planOrState *backupModel) (stateRefreshed *backupModel, found bool, diags diag.Diagnostics) {
+	stateRefreshed = planOrState.Copy()
+
+	var policiesFromAPI *v1.ProtectionLevels // *netbackupclient.ProtectionLevels
```
[the remark this answers is not in the document]
Comment (dmicheneau) [reply internal/provider/backup/backup_resource.go:413]: Fix

File: internal/provider/backup/backup_schema.go [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+			},
+			"target_id": superschema.SuperStringAttribute{
+				Common: &schemaR.StringAttribute{
+					MarkdownDescription: "The ID of the target. A target can be a VDC, a VApp or a VM.",
+					Optional:            true,
+				},
+				Resource: &schemaR.StringAttribute{
+					Validators: []validator.String{
+						stringvalidator.ExactlyOneOf(path.MatchRoot("target_id"), path.MatchRoot("target_name")),
+						fstringvalidator.IsUUID(),
```
[the remark this answers is not in the document]
Comment (dmicheneau) [reply internal/provider/backup/backup_schema.go]: already Fix

File: internal/provider/backup/backup_schema_test.go [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+
+	// The fwresource import alias is so there is no collistion
+	// with the more typical acceptance testing import:
+	// "github.com/hashicorp/terraform-plugin-testing/helper/resource".
+	fwresource "github.com/hashicorp/terraform-plugin-framework/resource"
+
+	netbackup "github.com/orange-cloudavenue/terraform-provider-cloudavenue/internal/provider/backup"
+)
+
+// TODO : Comment or uncomment the following imports if you are using resources or/and datasources
```
[the remark this answers is not in the document]
Comment (dmicheneau) [reply internal/provider/backup/backup_schema_test.go:16]: fix

File: templates/resources/backup.md.tmpl (hunk 7 above)
Comment (dmicheneau) [reply templates/resources/backup.md.tmpl:24]: Fix

File: internal/provider/backup/backup_schema.go (hunk 9 above)
Comment (dmicheneau) [reply internal/provider/backup/backup_schema.go:55]: Fix

File: internal/provider/backup/backup_schema.go (hunk 10 above)
Comment (dmicheneau) [reply internal/provider/backup/backup_schema.go:67]: Fix

File: internal/client/client.go (hunk 18 above)
Comment (dmicheneau) [reply internal/client/client.go:83]: Fix

**Review by azrod — APPROVED** [review]

**Pull request merged**